allennlp/data/iterators/basic_iterator.py

```diff
@@ -1,24 +1,52 @@
-from collections import deque
-from typing import Iterable, Deque
+from typing import Tuple, Iterable, Deque
 import logging
+from collections import deque
 import random
 
+from allennlp.data.iterators.data_iterator import DataIterator
+from allennlp.data.iterators.bucket_iterator import BucketIteratorShim
 from allennlp.common.util import lazy_groups_of
 from allennlp.data.instance import Instance
-from allennlp.data.iterators.data_iterator import DataIterator
 from allennlp.data.dataset import Batch
 
 logger = logging.getLogger(__name__)
 
 
 @DataIterator.register("basic")
-class BasicIterator(DataIterator):
+class BasicIteratorShim(BucketIteratorShim):
     """
     A very basic iterator that takes a dataset, possibly shuffles it, and creates fixed sized batches.
 
     It takes the same parameters as :class:`allennlp.data.iterators.DataIterator`
     """
 
+    def __init__(
+        self,
+        batch_size: int = 32,
+        instances_per_epoch: int = None,
+        max_instances_in_memory: int = None,
+        cache_instances: bool = False,
+        track_epoch: bool = False,
+        maximum_samples_per_batch: Tuple[str, int] = None,
+    ):
+
+        super().__init__(
+            batch_size=batch_size,
+            instances_per_epoch=instances_per_epoch,
+            max_instances_in_memory=max_instances_in_memory,
+            cache_instances=cache_instances,
+            track_epoch=track_epoch,
+            maximum_samples_per_batch=maximum_samples_per_batch,
+        )
+
+
+@DataIterator.register("basic_old")
+class BasicIterator(DataIterator):
```

Comment:
I'm in favour of stripping these old ones out, but it would break backward compatibility for anyone who has subclassed the iterators.

Reply:
I vote for just removing them. The exposure is probably small, and the reduction in complexity those people would gain by upgrading is worth it.

Reply:
@matt-gardner, I've been struggling with statements like "The exposure is probably small". It seems very difficult to actually measure this. At the same time, "reduction in complexity" is quite nebulous while "the code still runs" is very concrete. Given that we've been able to engineer this PR such that standard users of the jsonnet API will get the desired benefit it seems a shame to drop support for the more advanced subclass use-case.

Reply:
I guess I don't think we should be total slaves to backwards compatibility, at the expense of good, clean code. I agree with you on trying really hard to preserve compatibility on most releases, but we have an opportunity with 1.0 to significantly clean things up; let's not hamstring ourselves in the name of backwards compatibility.

Reply:
I agree with Matt here - I don't think the reduction in complexity is nebulous - it allows us to make changes to the codebase significantly faster. The diff on this PR would be half the size (in terms of lines added) if we removed the old iterators.

Additionally, it's at least clear in my mind that backward compatibility itself doesn't drive usage - look at pytorch transformers. 

Regardless, what we decide here makes a big difference to how I proceed with other parts of this PR, so...

Reply:
I'm very confused by the claim that the PR would be half the size in terms of lines added. Are you talking about not having the shims at all? If so, I think that would be a major mistake. You did a great job on the shims and I think it will be very helpful to maintain that for people.

The only additional lines of code to maintain the old iterators as-is is some duplicated comments and function/class prototypes. Maybe the tests would have been modified instead of duplicated, but that the work to upgrade them would be the same, so I'm not sure about moving faster...

Reply:
No, it's not about the shims - we need those for backwards compatibility of config files (which I definitely approve of).  It's about retaining the old classes themselves, registered as `"old_*"`.  I don't think we need to do that, the classes should just be removed.

What's the worst case that happens to someone who has subclassed one of the classes we remove?  They copy the code from an earlier release of allennlp into their repo, change an import line or two, and everything still works.  I don't think that's a very bad outcome, and it saves us a significant maintenance and code complexity burden.  Backwards compatibility at all costs is far too extreme; this is a case where the costs are worth breaking compatibility, in my opinion, as the costs are low and the gain is high.

Reply:
Absolutely agreed that backwards compatibility at all costs is too extreme. My claim is that the ongoing maintenance costs will actually be very low. Essentially we'd just slap a deprecation warning on the classes, a comment saying that users should use `TransformIterator` and never look at them again.

To clarify one point about the impact, this change doesn't just affect users that subclassed our iterators, it also affects those that wrote new iterator classes and dispatched to `_create_batches` (as its contract has changed). For instance, https://github.com/allenai/ernie/blob/master/ernie/self_attn_bucket_iterator.py.

At any rate, if we want to break backwards compatibility now is a good time, so consider my objection lodged and carry on if that's your decision.